src/imitation/algorithms/preference_comparisons.py

```diff
@@ -201,24 +207,42 @@ def __init__(
                 how good a fragment is. Default is to use undiscounted
                 sums of rewards (as in the DRLHP paper).
             seed: seed for the internal RNG (only used if temperature > 0)
+            threshold: preferences are sampled from a softmax of returns.
+                To avoid overflows, we clip differences in returns that are
+                above this threshold (after multiplying with temperature).
+                This threshold is therefore in logspace. The default value
+                of 50 means that probabilities below 2e-22 are rounded to 0.
+            custom_logger: Where to log to; if None (default), creates a new logger.
         """
         # we don't pass a logger for now since this particular implementation
         # doesn't use one at the moment
-        super().__init__()
+        super().__init__(custom_logger=custom_logger)
         self.temperature = temperature
         self.discount_factor = discount_factor
         self.rng = np.random.default_rng(seed=seed)
+        self.threshold = threshold
 
     def __call__(self, fragment_pairs: Sequence[TrajectoryWithRewPair]) -> np.ndarray:
-        rews1, rews2 = self._reward_sums(fragment_pairs)
+        returns1, returns2 = self._reward_sums(fragment_pairs)
         if self.temperature == 0:
-            return (np.sign(rews1 - rews2) + 1) / 2
+            return (np.sign(returns1 - returns2) + 1) / 2
+
+        returns1 /= self.temperature
+        returns2 /= self.temperature
 
-        rews1 /= self.temperature
-        rews2 /= self.temperature
+        # clip the returns to avoid overflows in the softmax below
+        returns_diff = np.clip(returns1 - returns2, -self.threshold, self.threshold)
         # Instead of computing exp(rews1) / (exp(rews1) + exp(rews2)) directly,
         # we divide enumerator and denominator by exp(rews1) to prevent overflows:
-        model_probs = 1 / (1 + np.exp(rews2 - rews1))
+        model_probs = 1 / (1 + np.exp(-returns_diff))
+        # Compute the mean binary entropy. This metric helps estimate
+        # how good we can expect the performance of the learned reward
+        # model to be at predicting preferences.
+        entropy = -(
+            special.xlogy(model_probs, model_probs)
+            + special.xlogy(1 - model_probs, 1 - model_probs)
+        ).mean()
```

Comment:
`sum()` not mean as `model_probs` are, well, probability weightings?

Reply:
I think just `scipy.stats.entropy(model_probs)` would do what you want.

Reply:
I thin `sum` is correct and that `scipy.stats.entropy(model_probs)` does something different. What might be causing some confusion: each entry of `model_probs` is a complete description of a single Bernoulli distribution (namely for one fragment pair). I want to compute the entropy of each Bernoulli distribution and then take the mean of all of them. This is analogous to how the `binary_cross_entropy` loss I'm using in `CrossEntropyTrainer` works, so if I have both metrics, I can tell how well the model does (because the mean entropy I'm computing is the lower bound on the mean cross-entropy).

Using `sum` or `mean` is pretty arbitrary but `mean` makes it more interpretable because it's independent from the batch size (and it's what pytorch uses for `binary_cross_entropy` by default.

I could use `scipy.stats.entropy`, but I'd have to stack together `model_probs` and `1 - model_probs` because `scipy`.stats.entropy` works on arbitrary categorical distributions rather than Bernoulli probabilities.

Reply:
OK, I'm convinced by this.